sql/distinct.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// distinct constructs a distinctNode.
+func (*planner) distinct(n *parser.Select) (*distinctNode, error) {
+	distinct := distinctNode{}
+	if n.Distinct {
+		distinct.seen = make(map[string]struct{})
+	}
+	return &distinct, nil
+}
+
+type distinctNode struct {
+	planNode
+	// All the columns that are part of the Sort. Set to nil if no-sort, or
+	// sort used an expression that was not part of the requested column set.
+	columnsInOrder []bool
+	// encoding of the columnsInOrder columns for the previous row.
+	lastSeen string
+	// encodings of the non-columnInOrder columns for rows sharing the same
+	// lastSeen value.
+	seen map[string]struct{}
+}
+
+// wrap the supplied planNode with the distinctNode if distinct is required.
+func (n *distinctNode) wrap(plan planNode) planNode {
+	if n.seen == nil {
+		return plan
+	}
+	n.planNode = plan
+	if len(plan.Ordering()) > 0 {
+		n.columnsInOrder = make([]bool, len(plan.Columns()), len(plan.Columns()))
+	}
+	for _, p := range plan.Ordering() {
```

Comment:
@petermattis do you think the tests are sufficient to assuage your fears that we should not be using Ordering(). See the ones what use SELECT y AS w.